src/sys/stm_tests.ml

```diff
@@ -0,0 +1,252 @@
+
+  (* val mem : filesys -> string list -> bool *)
+  let mem fs path = 
+    match find_opt fs path with
+    | None   -> false
+    | Some _ -> true
+
+  (* val mkdir : filesys -> string list -> string -> int -> filesys *)
+  let rec mkdir fs path new_dir_name perm = 
+    match fs with
+    | File _      -> fs
+    | Directory d -> 
+      (match path with 
+      | [] ->  
+        let new_dir = Directory {perm; fs_map = Map_names.empty} in
+        Directory {d with fs_map = Map_names.add new_dir_name new_dir d.fs_map}
+      | next_in_path :: tl_path -> 
+        (match Map_names.find_opt next_in_path d.fs_map with
+        | None        -> fs
+        | Some sub_fs -> 
+          let nfs = mkdir sub_fs tl_path new_dir_name perm in
+          if nfs = sub_fs
+          then fs
+          else 
+            let new_map = Map_names.remove next_in_path d.fs_map in
+            let new_map = Map_names.add next_in_path nfs new_map in
+            Directory {d with fs_map = new_map}))
+
+  (* val readdir : filesys -> string list -> string list option *)
+  let readdir fs path = 
+    match find_opt fs path with
+    | None    -> None
+    | Some fs -> 
+      match fs with
+      | File _ -> None
+      | Directory d -> Some (Map_names.fold (fun k _ l -> l @ [k]) d.fs_map [])
+
+
+  (* val rmdir : filesys -> string list -> string -> filesys *)
+  let rec rmdir fs path delete_dir_name = 
+    match fs with
+    | File _      -> fs
+    | Directory d -> 
+      (match path with 
+      | [] ->  
+        let target_fs = Map_names.find_opt delete_dir_name d.fs_map in
+        (match target_fs with
+        | None -> Directory d
+        | Some fs ->
+          (match fs with
+          | File _ -> Directory d
+          | Directory target -> 
+            let readd_target = (Map_names.fold (fun k _ l -> l @ [k]) target.fs_map []) in
+            if readd_target = []
+            then Directory {d with fs_map = Map_names.remove delete_dir_name d.fs_map}
+            else Directory d))
+      | next_in_path :: tl_path -> 
+        (match Map_names.find_opt next_in_path d.fs_map with
+        | None        -> fs
+        | Some sub_fs -> 
+          let nfs = rmdir sub_fs tl_path delete_dir_name in
+          if nfs = sub_fs
+          then fs
+          else 
+            let new_map = Map_names.remove next_in_path d.fs_map in
+            let new_map = Map_names.add next_in_path nfs new_map in
+            Directory {d with fs_map = new_map}))
+
+  let next_state c fs = 
+    match c with
+    | File_exists (_path) -> fs
+    | Mkdir (path, new_dir_name, perm) -> 
+      if mem fs (path @ [new_dir_name])
+      then fs
+      else mkdir fs path new_dir_name perm
+    | Rmdir (path,delete_dir_name) ->
+      if mem fs (path @ [delete_dir_name])
+      then rmdir fs path delete_dir_name
+      else fs
+    | Readdir _path -> fs 
+
+  let init_sut () = try Sys.mkdir (static_path / "sandbox_root") 0o777 with Sys_error _ -> ()
```

Comment:
Isn't it paradoxical to use `Sys.mkdir` at the very beginning of the test, whereas we are testing `Sys.mkdir` ?
Maybe it is better to call the `Sys.command` like in the 'cleanup function'.

Reply:
Yes - good question! That indeed requires `Sys.mkdir` to work reasonably :smile: 

`Sys.command` could work for now. On Windows `Sys.command` is going to work less well - but we can patch that later.